### PR TITLE
tls_process_server_hello: Disallow repeated HRR

### LINKED ISSUE #17934: Client allows more than one HelloRetryRequest
I previously reported this to the OpenSSL security address over concerns regarding execution time.

With the following Python script an input for the client fuzzer can be generated, which causes the client to invoke the Diffie-Hellman keygen many times over.

(use Python 2, not Python 3)
```python
hello = "\x16\x03\x0b\x00\x58\x02\x00\x00\x54\x03\x03\xcf\x21\xad\x74\xe5\x9a\x61\x11\xbe\x1d\x8c\x02\x1e\x65\xb8\x91\xc2\xa2\x11\x16\x7a\xbb\x8c\x5e\x07\x9e\x09\xe2\xc8\xa8\x33\x9c\x20\x01\x02\x03\x04\x05\x06\x07\x08\x09\x0a\x0b\x0c\x0d\x0e\x0f\x10\x11\x12\x13\x14\x15\x16\x17\x18\x19\x1a\x1b\x1c\x1d\x1e\x1f\x20\x13\x01\x00\x00\x0c\x00\x2b\x00\x02\x03\x04\x00\x33\x00\x02\x01\x03\x16\x03\x0b\x00\x58\x02\x00\x00\x54\x03\x03\xcf\x21\xad\x74\xe5\x9a\x61\x11\xbe\x1d\x8c\x02\x1e\x65\xb8\x91\xc2\xa2\x11\x16\x7a\xbb\x8c\x5e\x07\x9e\x09\xe2\xc8\xa8\x33\x9c\x20\x01\x02\x03\x04\x05\x06\x07\x08\x09\x0a\x0b\x0c\x0d\x0e\x0f\x10\x11\x12\x13\x14\x15\x16\x17\x18\x19\x1a\x1b\x1c\x1d\x1e\x1f\x20\x13\x01\x00\x00\x0c\x00\x2b\x00\x02\x03\x04\x00\x33\x00\x02\x01\x04"
header = "\x16\x03\x0b\x00\x58\x02\x00\x00\x54\x03\x03\xcf\x21\xad\x74\xe5\x9a\x61\x11\xbe\x1d\x8c\x02\x1e\x65\xb8\x91\xc2\xa2\x11\x16\x7a\xbb\x8c\x5e\x07\x9e\x09\xe2\xc8\xa8\x33\x9c\x20\x01\x02\x03\x04\x05\x06\x07\x08\x09\x0a\x0b\x0c\x0d\x0e\x0f\x10\x11\x12\x13\x14\x15\x16\x17\x18\x19\x1a\x1b\x1c\x1d\x1e\x1f\x20\x13\x01\x00\x00\x0c\x00\x2b\x00\x02\x03\x04\x00\x33\x00\x02\x01\x04\x16\x03\x0b\x00\x58\x02\x00\x00\x54\x03\x03\xcf\x21\xad\x74\xe5\x9a\x61\x11\xbe\x1d\x8c\x02\x1e\x65\xb8\x91\xc2\xa2\x11\x16\x7a\xbb\x8c\x5e\x07\x9e\x09\xe2\xc8\xa8\x33\x9c\x20\x01\x02\x03\x04\x05\x06\x07\x08\x09\x0a\x0b\x0c\x0d\x0e\x0f\x10\x11\x12\x13\x14\x15\x16\x17\x18\x19\x1a\x1b\x1c\x1d\x1e\x1f\x20\x13\x01\x00\x00\x0c\x00\x2b\x00\x02\x03\x04\x00\x33\x00\x02\x01\x03\x16\x03\x0b\x00\x58\x02\x00\x00\x54\x03\x03\xcf\x21\xad\x74\xe5\x9a\x61\x11\xbe\x1d\x8c\x02\x1e\x65\xb8\x91\xc2\xa2\x11\x16\x7a\xbb\x8c\x5e\x07\x9e\x09\xe2\xc8\xa8\x33\x9c\x20\x01\x02\x03\x04\x05\x06\x07\x08\x09\x0a\x0b\x0c\x0d\x0e\x0f\x10\x11\x12\x13\x14\x15\x16\x17\x18\x19\x1a\x1b\x1c\x1d\x1e\x1f\x20\x13\x01\x00\x00\x0c\x00\x2b\x00\x02\x03\x04\x00\x33\x00\x02\x01\x04\x16\x03\x0b\x00\x58\x02\x00\x00\x54\x03\x03\xcf\x21\xad\x74\xe5\x9a\x61\x11\xbe\x1d\x8c\x02\x1e\x65\xb8\x91\xc2\xa2\x11\x16\x7a\xbb\x8c\x5e\x07\x9e\x09\xe2\xc8\xa8\x33\x9c\x20\x01\x02\x03\x04\x05\x06\x07\x08\x09\x0a\x0b\x0c\x0d\x0e\x0f\x10\x11\x12\x13\x14\x15\x16\x17\x18\x19\x1a\x1b\x1c\x1d\x1e\x1f\x20\x13\x01\x00\x00\x0c\x00\x2b\x00\x02\x03\x04\x00\x33\x00\x02\x01\x03\x16\x03\x0b\x00\x58\x02\x00\x00\x54\x03\x03\xcf\x21\xad\x74\xe5\x9a\x61\x11\xbe\x1d\x8c\x02\x1e\x65\xb8\x91\xc2\xa2\x11\x16\x7a\xbb\x8c\x5e\x07\x9e\x09\xe2\xc8\xa8\x33\x9c\x20\x01\x02\x03\x04\x05\x06\x07\x08\x09\x0a\x0b\x0c\x0d\x0e\x0f\x10\x11\x12\x13\x14\x15\x16\x17\x18\x19\x1a\x1b\x1c\x1d\x1e\x1f\x20\x13\x01\x00\x00\x0c\x00\x2b\x00\x02\x03\x04\x00\x33\x00\x02\x01\x04\x16\x03\x0b\x00\x58\x02\x00\x00\x54\x03\x03\xcf\x21\xad\x74\xe5\x9a\x61\x11\xbe\x1d\x8c\x02\x1e\x65\xb8\x91\xc2\xa2\x11\x16\x7a\xbb\x8c\x5e\x07\x9e\x09\xe2\xc8\xa8\x33\x9c\x20\x01\x02\x03\x04\x05\x06\x07\x08\x09\x0a\x0b\x0c\x0d\x0e\x0f\x10\x11\x12\x13\x14\x15\x16\x17\x18\x19\x1a\x1b\x1c\x1d\x1e\x1f\x20\x13\x01\x00\x00\x0c\x00\x2b\x00\x02\x03\x04\x00\x33\x00\x02\x01\x03\x16\x03\x0b\x00\x58\x02\x00\x00\x54\x03\x03\xcf\x21\xad\x74\xe5\x9a\x61\x11\xbe\x1d\x8c\x02\x1e\x65\xb8\x91\xc2\xa2\x11\x16\x7a\xbb\x8c\x5e\x07\x9e\x09\xe2\xc8\xa8\x33\x9c\x20\x01\x02\x03\x04\x05\x06\x07\x08\x09\x0a\x0b\x0c\x0d\x0e\x0f\x10\x11\x12\x13\x14\x15\x16\x17\x18\x19\x1a\x1b\x1c\x1d\x1e\x1f\x20\x13\x01\x00\x00\x0c\x00\x2b\x00\x02\x03\x04\x00\x33\x00\x02\x01\x04"

out = header
for i in range(1000):
    out += hello
with open('x.bin', 'wb') as fp:
    fp.write(out)
```

I was informed by @davidben that the actual bug here is the fact that the client is responding to more than 1 HelloRetryRequest from the server.

I think it's this:

https://datatracker.ietf.org/doc/html/rfc8446#section-4.1.4:

```
If a client receives a second
HelloRetryRequest in the same connection (i.e., where the ClientHello
was itself in response to a HelloRetryRequest), it MUST abort the
handshake with an "unexpected_message" alert
```

### PR DESCRIPTION
**Pull request by t8m**
Repeated HRR must be rejected.

Fixes #17934

##### Checklist

- [x] tests are added or updated

### CODE REVIEW

**Comment by t8m** [comment]
I'd like to add a test.

@mattcaswell Is there a way how to test this via TLS Proxy?

Otherwise I'd just do something with the fuzz data and create a test code similar to the client fuzzer.

**Comment by mattcaswell** [comment]
> @mattcaswell Is there a way how to test this via TLS Proxy?

Yes. Here is an example of a TLSProxy test injecting a custom ClientHello into the communication. That particular test injects it in an SSLv2 record - but that's a simple flag that you wouldn't use in your test. The same principles should be applicable to injecting an extra HRR.

https://github.com/openssl/openssl/blob/9e1a54f4a187195fc417ad0f90e84d208d478968/test/recipes/70-test_sslrecords.t#L336-L406

**Comment by mattcaswell** [comment]
> @mattcaswell do you have any idea why the test fails? Basically the alert is sent by the client much later than expected - in flight 4 and not in flight 2.

I will try and take a look later today.

**Comment by t8m** [comment]
It is interesting that sometimes it is sent in flight 2 and sometimes in flight 4. Perhaps depends on timing?

**Review by tmshort — APPROVED** [review]
Seems reasonable, but yeah, would like to understand the flight number inconsistency.

File: test/recipes/70-test_tls13hrr.t [hunk 1]
```diff
@@ -35,9 +35,12 @@ my $proxy = TLSProxy::Proxy->new(
 
 use constant {
     CHANGE_HRR_CIPHERSUITE => 0,
-    CHANGE_CH1_CIPHERSUITE => 1
+    CHANGE_CH1_CIPHERSUITE => 1,
+    DUPLICATE_HRR => 2
 };
 
+plan tests => 3;
```
[the remark this answers is not in the document]
Comment (t8m) [reply test/recipes/70-test_tls13hrr.t:42]: It was (see below) - I moved the statement to more suitable place.

[the remark this answers is not in the document]
Comment (tmshort) [reply test/recipes/70-test_tls13hrr.t:42]: Ah yes... I've seen the odd placement of that in other files, too...

[the remark this answers is not in the document]
Comment (tmshort) [reply test/recipes/70-test_tls13hrr.t:42]: (I approved the C code; as Matt was going to look at why the test was behaving odd)

**Comment by mattcaswell** [comment]
Flight 4 seems correct?

````
Client                                        Server
-------                                       ------
Initial Client Hello (flight 0)
                                              HRR1 (flight 1)
Second Client Hello (flight 2)
                                              HRR2 (flight 3)
Alert (flight 4)
````

But flight 2 does not seem right.

**Comment by mattcaswell** [comment]
Ping? What's happening with this?

**Comment by t8m** [comment]
Yeah, this looks better now. @tmshort please reconfirm.

**Review by tmshort — APPROVED** [review]
LGTM

**Review by mattcaswell — APPROVED** [review]

**Comment by t8m** [comment]
Merged to master, 3.0, and 1.1.1 branches. Thank you for your reviews.

**Comment by ShlomobCohen** [comment]
Need to include:

#include "statem.h"

for SSLfatal macro usage.

**Pull request closed without merging**